src/devices/bus/ieee488/remote488.cpp

```diff
@@ -0,0 +1,721 @@
+		m_connect_cnt = MAX_MISSED_HB;
+	} else {
+		if (m_connected) {
+			// Disconnected
+			m_connected = false;
+			LOG("Connection lost!\n");
+			update_ah_fsm();
+			m_hb_timer->adjust(attotime::from_msec(HEARTBEAT_MS) , 0 , attotime::from_msec(HEARTBEAT_MS));
+		}
+	}
+}
+
+void remote488_device::recvd_data_byte(uint8_t data , bool eoi)
+{
+	m_ib = data;
+	m_ib_eoi = eoi;
+	m_ibf = true;
+	update_sh_fsm();
+	update_ah_fsm();
+}
+
+void remote488_device::update_signals_from_rem(uint8_t to_set , uint8_t to_clear)
+{
+	uint8_t diff = m_in_signals;
+	m_in_signals |= to_set;
+	m_in_signals &= ~to_clear;
+	diff ^= m_in_signals;
+	m_out_signals = m_in_signals;
+
+	//LOG("REM SIG %02x %02x\n" , m_in_signals , diff);
+	m_no_propagation = true;
+	uint8_t tmp = m_out_signals;
+
+	if (BIT(diff , SIGNAL_ATN_BIT)) {
+		m_bus->atn_w(this , BIT(m_in_signals , SIGNAL_ATN_BIT));
+		COPY_BIT(m_bus->atn_r() , tmp , SIGNAL_ATN_BIT);
+	}
+	if (BIT(diff , SIGNAL_IFC_BIT)) {
+		m_bus->ifc_w(this , BIT(m_in_signals , SIGNAL_IFC_BIT));
+		COPY_BIT(m_bus->ifc_r() , tmp , SIGNAL_IFC_BIT);
+	}
+	if (BIT(diff , SIGNAL_REN_BIT)) {
+		m_bus->ren_w(this , BIT(m_in_signals , SIGNAL_REN_BIT));
+		COPY_BIT(m_bus->ren_r() , tmp , SIGNAL_REN_BIT);
+	}
+	if (BIT(diff , SIGNAL_SRQ_BIT)) {
+		m_bus->srq_w(this , BIT(m_in_signals , SIGNAL_SRQ_BIT));
+		COPY_BIT(m_bus->srq_r() , tmp , SIGNAL_SRQ_BIT);
+	}
+
+	m_no_propagation = false;
+
+	update_state(tmp);
+}
+
+void remote488_device::update_signal(signal_bit bit , int state)
+{
+	if (!m_no_propagation) {
+		uint8_t tmp = m_out_signals;
+		COPY_BIT(state , tmp , bit);
+		update_state(tmp);
+	}
+}
+
+void remote488_device::update_state(uint8_t new_signals)
+{
+	uint8_t to_set = new_signals & ~m_out_signals;
+	uint8_t to_clear = ~new_signals & m_out_signals;
+
+	m_out_signals = new_signals;
+
+	if (is_local_atn_active() && m_ibf) {
+		LOG("Flushing enabled\n");
+		m_flush_bytes = true;
+		m_ibf = false;
+		m_poll_timer->adjust(attotime::zero);
+	}
+
+	if (to_set) {
+		send_update(MSG_SIGNAL_SET , to_set);
+	}
+	if (to_clear) {
+		send_update(MSG_SIGNAL_CLEAR , to_clear);
+	}
+}
+
+void remote488_device::send_update(char type , uint8_t data)
+{
+	std::ostringstream buff;
+	util::stream_format(buff , "%c:%02x\n" , type , data);
+	LOG("%.6f %s" , machine().time().as_double() , buff.str().c_str());
+	std::string out{buff.str()};
+	for (auto c : out) {
+			m_stream->output(c);
+	}
+}
+
+bool remote488_device::a2hex(char c , uint8_t& out)
+{
+	if (c >= '0' && c <= '9') {
+		out = c - '0';
+		return true;
+	} else if (c >= 'a' && c <= 'f') {
+		out = c - 'a' + 10;
+		return true;
+	} else if (c >= 'A' && c <= 'F') {
+		out = c - 'A' + 10;
+		return true;
+	} else {
+		return false;
+	}
+}
+
+char remote488_device::recv_update(uint8_t& data)
+{
+	char c;
+	unsigned i;
+
+	// Do not iterate too much..
+	for (i = 0; i < 8 && m_stream->input(&c , 1); i++) {
+		switch (m_rx_state) {
+		case REM_RX_WAIT_CH:
+			if (isalnum(c)) {
```

Comment:
Functions like `isalnum` are locale-aware.  It's generally a bad idea to use them in situations like this.